### PR TITLE
Add bundles

### PR DESCRIPTION
**Pull request by riasvdv**

### CODE REVIEW

File: app/Actions/HandlePurchaseAction.php [hunk 1]
```diff
@@ -66,29 +70,38 @@ public function execute(
         }
 
         if ($user->email) {
-            Mail::to($user->email)->queue(new PurchaseConfirmationMail($purchase));
+            foreach ($purchasables as $purchasable) {
+                Mail::to($user->email)->queue(new PurchaseConfirmationMail($purchase, $purchasable));
```
Comment (riasvdv) [inline app/Actions/HandlePurchaseAction.php:74]: Should we send a custom "bundle" purchase confirmation email?

Comment (freekmurze) [reply app/Actions/HandlePurchaseAction.php:74]: Let's keep it simpel and send the mails of the separate products inside the bundle.

File: app/Http/Controllers/AfterPaddleBundleSaleController.php [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+        session()->flash('sold_purchasable', $bundle->purchasables->first());
+
+        if (current_user()) {
+            session()->flash('latest_purchase', current_user()->purchases()->latest()->first());
+        }
+
+        flash()->success('Purchase successful!');
+
+        return redirect()->route('products.show', $bundle->purchasables->first()->product);
```
Comment (riasvdv) [inline app/Http/Controllers/AfterPaddleBundleSaleController.php:25]: We redirect to the first product inside the bundle after the sale here. As bundles don't really have a dedicated detail page right now (I opted to skip this and go straight to the "buy" page).

We could redirect to the user's purchases instead?

Comment (freekmurze) [reply app/Http/Controllers/AfterPaddleBundleSaleController.php:25]: Redirecting to the user's purchases seems like a good solution. A flash message should indicate that new products have been added.

File: app/Actions/HandlePurchaseAction.php [hunk 3]
```diff
@@ -41,16 +42,19 @@ public function __construct(
 
     public function execute(
         User $user,
-        Purchasable $purchasable,
+        Bundle|Purchasable $purchasable,
```
Comment (freekmurze) [inline app/Actions/HandlePurchaseAction.php:45]: Nice!

File: app/Models/Referrer.php [hunk 4]
```diff
@@ -68,9 +73,9 @@ public function getDiscountPercentage(Purchasable $purchasable): int
         return 0;
     }
 
-    public function hasActiveDiscount(Purchasable $purchasable): bool
+    public function hasActiveDiscount(Model $model): bool
```
Comment (freekmurze) [inline app/Models/Referrer.php:76]: Could we typehint specific models here?

File: app/Models/User.php [hunk 5]
```diff
@@ -147,7 +173,7 @@ public function purchasesWithoutRenewals(): HasMany
                 PurchasableType::TYPE_STANDARD_RENEWAL,
                 PurchasableType::TYPE_UNLIMITED_DOMAINS_RENEWAL,
             ]);
-        });
+        })->orWhereHas('bundle');
```
Comment (freekmurze) [inline app/Models/User.php:176]: Is this correct? Seems like all bundle buyers will be regarded as sponsors.

**Review by freekmurze — CHANGES_REQUESTED** [review]
Seems good, after the comments have been addressed, this may be merged and deployed

File: app/Models/User.php (hunk 5 above)
Comment (riasvdv) [reply app/Models/User.php:176]: This query doesn't have anything to do with sponsors? It retrieves all purchases without any renewals. So bundle purchases should be included

Comment (freekmurze) [reply app/Models/User.php:176]: Right! I was confused by the diff 🙂

**Pull request merged**